streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -2433,8 +2421,41 @@ object ZChannel {
               ZChannel.write(outElem) *> readerCh(totalStreams, completedStreams, accRes)
           }
 
+      // when unbounded, we run the risk of spawning a fiber per channel,
+      // even worse, each such fiber may end up processing one or less channels...
+      // this attempts to mitigate this by
+      // 1. detecting scenarios where an idle worker was able to immediately pick the enqueued message, as a result we get a 'best effort' behavior of limiting the number of fibers
+      // 2. shutting down fibers when there is no more work to be done. Once a fiber processes input, it optimistically tries to fetch another value to process.
+      //    if there's nothing to fetch, it stops pulling data and notifies this coordinator which may decide to resurrect it
+      // The coordinator is basically a counter with linearized access to the worker pool,
+      // and is needed to avoid races between spawning workers on element pull and shutting them down on empty queue
+      val coordinatorFiber =
+        ZStream
+          .fromQueueWithShutdown(workerPoolCoordinator)
+          .runFoldScopedZIO(0) { case (runningFibers, requestNewWorker) =>
```

Comment:
Could be done with raw Channels as everything else here for efficiency. It was a POC to see if it'll at least work